src/compile/scale/domain.ts

```diff
@@ -162,3 +162,67 @@ function _useRawDomain (scale: Scale, model: Model, channel: Channel) {
       (fieldDef.type === 'temporal' && util.contains([ScaleType.TIME, ScaleType.UTC], scale.type))
     );
 }
+
+
+/**
+ * Convert the domain to an array of data refs or an array of values. Also, throw
+ * away sorting information since we always sort the domain when we union two domains.
+ */
+function normalizeDomain(domain: VgDomain): (any[] | VgDataRef)[] {
+  if (util.isArray(domain)) {
+    return [domain];
+  } if (isDataRefDomain(domain)) {
+    delete domain.sort;
+    return [domain];
+  } else if(isFieldRefUnionDomain(domain)) {
+    return domain.fields.map(d => {
+      return {
+        data: domain.data,
+        field: d
+      };
+    });
+  } else if (isDataRefUnionedDomain(domain)) {
+    return domain.fields.map(d => {
+      if (util.isArray(d)) {
+        return d;
+      }
+      return {
+        field: d.field,
+        data: d.data
+      };
+    });
+  }
+
+  throw 'Should not get here.';
+}
+
+/**
+ * Union two data domains. A unioned domain is always sorted.
+ */
+export function unionDomains(domain1: VgDomain, domain2: VgDomain): VgDomain {
+  const normalizedDomain1 = normalizeDomain(domain1);
+  const normalizedDomain2 = normalizeDomain(domain2);
+
+  let fields = normalizedDomain1.concat(normalizedDomain2);
```

Comment:
I know this is `fields` in the Vega property, but it's not really an array of "field" but rather either an array of `domain array` or `DataRef`.  I wonder if `domains` is better?  (This is partly why I think `fields` is a pretty confusing property name.) 

I wonder if VSCode refactor would do this correct though.  Maybe You'll need to change `{fields, sort: true}` below to `{fields: domains, sort: true}`.  (If so, this could one problem for `{fields}` syntax.)